internal/controller/iprangeclaim_controller.go

```diff
@@ -0,0 +1,323 @@
+type IpRangeClaimReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	NetboxClient      *api.NetboxClient
+	Recorder          record.EventRecorder
+	OperatorNamespace string
+	RestConfig        *rest.Config
+}
+
+//+kubebuilder:rbac:groups=netbox.dev,resources=iprangeclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=netbox.dev,resources=iprangeclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=netbox.dev,resources=iprangeclaims/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *IpRangeClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("reconcile loop started")
+
+	o := &netboxv1.IpRangeClaim{}
+	err := r.Client.Get(ctx, req.NamespacedName, o)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if being deleted
+	if !o.ObjectMeta.DeletionTimestamp.IsZero() {
+		// end loop if deletion timestamp is not zero
+		return ctrl.Result{}, nil
+	}
+
+	//  check if matching IpRange object already exists
+	ipRange := &netboxv1.IpRange{}
+	ipRangeName := o.ObjectMeta.Name
+	ipRangeLookupKey := types.NamespacedName{
+		Name:      ipRangeName,
+		Namespace: o.Namespace,
+	}
+
+	err = r.Client.Get(ctx, ipRangeLookupKey, ipRange)
+	if err != nil {
+		// return error if not a notfound error
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		logger.V(4).Info("iprange object matching iprange claim was not found, creating new iprange object")
+
+		ipRangeModel, res, err := r.restoreOrAssignIpRangeAndSetCondition(ctx, o)
+		if ipRangeModel == nil {
+			return res, err
+		}
+
+		// create the IpRange CR
+		ipRangeResource := generateIpRangeFromIpRangeClaim(ctx, o, ipRangeModel.StartAddress, ipRangeModel.EndAddress)
+		err = controllerutil.SetControllerReference(o, ipRangeResource, r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = r.Client.Create(ctx, ipRangeResource)
+		if err != nil {
+			errSetCondition := r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedFalse, corev1.EventTypeWarning, "", err)
+			if errSetCondition != nil {
+				return ctrl.Result{}, errSetCondition
+			}
+			return ctrl.Result{}, err
+		}
+
+		err = r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedTrue, corev1.EventTypeNormal,
+			fmt.Sprintf(" , assigned ip range: %s-%s", ipRangeModel.StartAddress, ipRangeModel.EndAddress), nil)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+	} else {
+		// update fields of IPRange object
+		logger.V(4).Info("update iprange resource")
+		err := r.Client.Get(ctx, ipRangeLookupKey, ipRange)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Why is there a second `Get`? Nothing should have changed here as we are in the 'nil error' part of the branch, i.e. we got `ipRange`